torch/csrc/jit/codegen/cuda/index_compute.cpp

```diff
@@ -1937,52 +1937,51 @@ std::vector<Val*> Index::getNonGlobalProducerStridedIndices(
   return strided_inds;
 }
 
-std::vector<Val*> Index::getLinearIndex(
-    TensorView* consumer_tv,
-    const std::vector<kir::ForLoop*>& loops) {
+std::vector<Val*> Index::getRootIndicesWithContiguity(
+    TensorView* tv,
+    const std::vector<kir::ForLoop*>& loops,
+    bool contiguity) {
   // Use domain guard to ignore the contiguity of
   //  consumer tv.
-  TensorDomain* consumer_tv_no_contiguity_domain = nullptr;
-  auto contiguity_vector =
-      std::vector<bool>(consumer_tv->getMaybeRFactorDomain().size(), true);
-  if (consumer_tv->hasRFactor()) {
-    consumer_tv_no_contiguity_domain = IrBuilder::create<TensorDomain>(
-        consumer_tv->getRootDomain(),
-        consumer_tv->getRFactorDomain(),
-        consumer_tv->domain()->domain(),
+  TensorDomain* domain_with_specified_contiguity = nullptr;
+  std::vector<bool> contiguity_vector(
+      tv->getMaybeRFactorDomain().size(), contiguity);
+  if (tv->hasRFactor()) {
+    domain_with_specified_contiguity = IrBuilder::create<TensorDomain>(
+        tv->getRootDomain(),
+        tv->getRFactorDomain(),
+        tv->domain()->domain(),
         contiguity_vector);
   } else {
-    consumer_tv_no_contiguity_domain = IrBuilder::create<TensorDomain>(
-        consumer_tv->getRootDomain(),
-        consumer_tv->domain()->domain(),
-        contiguity_vector);
+    domain_with_specified_contiguity = IrBuilder::create<TensorDomain>(
+        tv->getRootDomain(), tv->domain()->domain(), contiguity_vector);
   }
 
-  ir_utils::TVDomainGuard domain_guard(
-      consumer_tv, consumer_tv_no_contiguity_domain);
+  ir_utils::TVDomainGuard domain_guard(tv, domain_with_specified_contiguity);
 
-  // TODO:
-  //  More optimization on the underlying tensor layout
-  //   will be done in a follow up.
-  return getGlobalConsumerStridedIndices(consumer_tv, loops);
+  IndexFromIdGraph index_from_id_graph = getTensorIndexFromIdGraph(loops, tv);
+  return getRootIndices(tv, loops, index_from_id_graph);
 }
 
-std::vector<Val*> Index::getGlobalConsumerStridedIndices(
-    const TensorView* consumer_tv,
+Val* Index::getLinearIndex(
+    TensorView* consumer_tv,
     const std::vector<kir::ForLoop*>& loops) {
-  FUSER_PERF_SCOPE("GpuLower::Lower::getGlobalConsumerIndex");
-
-  auto gpu_lower = GpuLower::current();
-
-  auto index_from_id_graph = getTensorIndexFromIdGraph(loops, consumer_tv);
+  return getRootIndicesWithContiguity(consumer_tv, loops, true).back();
```

Comment:
Should it be the sum of the the returned value instead of the last one? Not sure if the contiguous path would guarantee to return a single index term.

Reply:
The contiguous path returns [0, 0, 0, ..., linear index]

Reply:
I don't think that's a guaranteed behavior, at least I don't think that's an invariant that's actively being maintained.

My guess is there could be `split/merge` patterns that make one of the zero's non-zero with the underlying contig finder and zero propagation we have today. 

In other indexing interfaces we still try to manually multiply the strides and accumulate, which I think would be safer here too.

Reply:
😱 If it is not guaranteed, then should I switch back using `getGlobalConsumerStridedIndices`?

Reply:
Oh yeah, found an example to break this:
```C++
TEST_F(NVFuserTest, FusionRNGManualScheduleValidateWithCURand2_CUDA) {
  auto dtype = kFloat;
  std::unique_ptr<Fusion> fusion_ptr = std::make_unique<Fusion>();
  auto fusion = fusion_ptr.get();
  FusionGuard fg(fusion);

  Int* size1 = IrBuilder::create<Int>();
  Int* size2 = IrBuilder::create<Int>();
  Int* size3 = IrBuilder::create<Int>();
  Int* size4 = IrBuilder::create<Int>();
  fusion->addInput(size1);
  fusion->addInput(size2);
  fusion->addInput(size3);
  fusion->addInput(size4);
  TensorView* tv0 = rand({size1, size2, size3, size4}, DataType::Float);
  fusion->addOutput(tv0);

  auto options = at::TensorOptions().dtype(dtype).device(at::kCUDA, 0);

  FusionExecutor fe;
  fe.compileFusion(fusion, {10, 10, 10, 10});

  at::manual_seed(0);
  auto cg_outputs = fe.runFusion({10, 10, 10, 10});
  auto out = cg_outputs[0];

  at::manual_seed(0);
  auto ref = generate_uniform(10000, dtype).view({10, 10, 10, 10});

  testValidate(fusion, {out}, {10, 10, 10, 10}, {ref}, __LINE__, __FILE__);
}
```
generated wrong kernel
```CUDA
__global__ void kernel1(int64_t i0, int64_t i1, int64_t i2, int64_t i3, Tensor<float, 4> T0, at::PhiloxCudaState philox_args) {
  auto philox_offset = philox_args.captured_ ?
    static_cast<uint64_t>(*(philox_args.offset_.ptr) + philox_args.offset_intragraph_) :
    philox_args.offset_.val;
  uint4 rng_result;
  nvfuser_index_t rng_subseq = -1;
  nvfuser_index_t rng_offset = -1;
  #pragma unroll 1
  for(nvfuser_index_t i5 = 0; i5 < i0; ++i5) {
    #pragma unroll 1
    for(nvfuser_index_t i6 = 0; i6 < i1; ++i6) {
      #pragma unroll 1
      for(nvfuser_index_t i7 = 0; i7 < i2; ++i7) {
        #pragma unroll 1
        for(nvfuser_index_t i8 = 0; i8 < i3; ++i8) {
          nvfuser_index_t linear_index17 = i8;
          nvfuser_index_t rng_subseq17 = linear_index17 / 4;
          nvfuser_index_t rng_component17 = linear_index17 % 4;
          nvfuser_index_t rng_offset17 = 0;
          if (rng_subseq != rng_subseq17 || rng_offset != rng_offset17) {
            rng_result = philox(philox_args.seed_, rng_subseq17, philox_offset / 4 + rng_offset17);
            rng_subseq = rng_subseq17;
            rng_offset = rng_offset17;
          }
          T0[(i5 * ((i3 * i2) * i1)) + (i6 * (i3 * i2)) + (i7 * i3) + i8] = rng_uniformf(rng_result, rng_component17);
        }
      }
    }
  }
}
```
Thank you so much for pointing this out